include/revng/Model/TupleTree.h

```diff
@@ -1205,7 +1228,14 @@ class TupleTree {
     bool Result = true;
 
     visitReferences([&Result, this](const auto &Element) {
-      Result = Result and (Element.Root == Root.get());
+      const auto SameRoot = [&]() {
+        const auto GetPtrToConstRoot = [](const auto &RootPointer) {
+          return const_cast<const T *>(RootPointer);
```

Comment:
I don't think you need a `const_cast` to *add* const-ness.
Let's not raise unnecessary suspicion to the reader with a `const_cast`.

Reply:
If you don't explicitly use `const_cast` here, the `GetPtrToConstRoot` lambda will not have a unique return type for the various types of `RootPointer`. Here's why.

The generic lambda here, without the `const_cast`, expands to something like:
```c++
template<typename T>
auto GetPtrToConstRoot(const T& RootPointer) {
  return RootPointer;
}
```

Given that we are using this to visit a `std::variant<RootT *, const Root *T>`, when substituting the two different types of the variant, it will return a `RootT *` in a case and a `const RootT *` in another case.

This does not work with `std::visit` because it requires the visitor to have a single return type. Hence I needed to explicitly add the `const_cast`, so that all the instantiations of the generic lambda on the various types in the variant will return the same type.

If I don't do that, the code does not compile:
```
error: static_assert failed due to requirement 'integral_constant<bool, false>::value' "`std::visit` requires the visitor to have a single return type."
```

Maybe there's a cleaner way out that does not require `const_cast`. If so, feel free to suggest it!

Reply:
I was just suggesting to use `static_cast`.
But you can also specify the return type of the lambda with something like `-> const decltype(RootPointer)`.
You can also make add an explicit template parameter list to the lambda if you prefer, but I'd just go with a `static_cast`.

Reply:
Oh, I see what you mean. I'm going for `-> const T *` then, which is more concise and exactly what we want.

Reply:
Done, take a look

Reply:
Looks great!